### PR TITLE
Convert transaction descriptions from English to python code

### PR DESCRIPTION
**Pull request by vbuterin**
Only a change of presentation, does not make any substantive changes.

Does NOT handle Deposit transactions because that is a special case for which I recommend a deeper change (will do this in another PR soon).

### CODE REVIEW

File: specs/core/0_beacon-chain.md [hunk 1]
```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
...
-* Verify that either (i) `state.latest_crosslinks[attestation.data.shard] == attestation.data.latest_crosslink` or (ii) `state.latest_crosslinks[attestation.data.shard] == Crosslink(crosslink_data_root=attestation.data.crosslink_data_root, epoch=slot_to_epoch(attestation.data.slot))`.
-* Verify bitfields and aggregate signature:
+For each `attestation` in `block.body.attestations`, , run the following function:
 
 ```python
-    assert attestation.custody_bitfield == b'\x00' * len(attestation.custody_bitfield)  # [TO BE REMOVED IN PHASE 1]
+def process_attestation(attestation: Attestation, state: BeaconState):
+    # Can't submit attestations that are too far in history (or in prehistory) 
+    assert attestation.data.slot >= GENESIS_SLOT
+    state.slot < attestation.data.slot + SLOTS_PER_EPOCH
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
    assert state.slot < attestation.data.slot + SLOTS_PER_EPOCH
```

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
...
+    # Can't submit attestations too quickly
+    attestation.data.slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
    assert attestation.data.slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot
```

File: specs/core/0_beacon-chain.md [hunk 3]
```diff
@@ -1741,32 +1795,66 @@ process_deposit(state, deposit)
...
 
 ##### Transfers
 
 Note: Transfers are a temporary functionality for phases 0 and 1, to be removed in phase 2.
 
 Verify that `len(block.body.transfers) <= MAX_TRANSFERS` and that all transfers are distinct.
 
-For each `transfer` in `block.body.transfers`:
-
-* Verify that `state.validator_balances[transfer.from] >= transfer.amount`.
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]: These first two statements were for overflow protection.

**Comment by JustinDrake** [comment]
Agreed with Danny, this is fantastic. Doing something similar for "Per-epoch processing" would be fab.

File: specs/core/0_beacon-chain.md [hunk 4]
```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
...
-* Verify that `proposer_slashing.proposal_1.block_root != proposer_slashing.proposal_2.block_root`.
-* Verify that `proposer.slashed is False`.
-* Verify that `bls_verify(pubkey=proposer.pubkey, message_hash=signed_root(proposer_slashing.proposal_1, "signature"), signature=proposer_slashing.proposal_1.signature, domain=get_domain(state.fork, slot_to_epoch(proposer_slashing.proposal_1.slot), DOMAIN_PROPOSAL))`.
-* Verify that `bls_verify(pubkey=proposer.pubkey, message_hash=signed_root(proposer_slashing.proposal_2, "signature"), signature=proposer_slashing.proposal_2.signature, domain=get_domain(state.fork, slot_to_epoch(proposer_slashing.proposal_2.slot), DOMAIN_PROPOSAL))`.
-* Run `slash_validator(state, proposer_slashing.proposer_index)`.
+For each `proposer_slashing` in `block.body.proposer_slashings`, run the following function:
+
+```python
+def process_proposer_slashing(proposer_slashing: ProposerSlashing,
+                              state: BeaconState):
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Prefer to re-order the parameters: `state` and then `proposer_slashing` as most helper functions.

File: specs/core/0_beacon-chain.md [hunk 5]
```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
...
+    assert (
+        is_double_vote(attestation1.data, attestation2.data) or
+        is_surround_vote(attestation1.data, attestation2.data)
+    )
+    assert verify_slashable_attestation(state, attestation1)
+    assert verify_slashable_attestation(state, attestation2)
+    slashable_indices = [
+        index for index in attestation1.validator_indices if
+        index in attestation2.validator_indices and state.validator_registry[index].slashed is False
+    ]
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Formatting :
```python
    slashable_indices = [
        index for index in attestation1.validator_indices
        if (
            index in attestation2.validator_indices and
            state.validator_registry[index].slashed is False
        )
    ]
```

File: specs/core/0_beacon-chain.md [hunk 6]
```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
...
-* Verify that `slashable_attestation_1.data != slashable_attestation_2.data`.
-* Verify that `is_double_vote(slashable_attestation_1.data, slashable_attestation_2.data)` or `is_surround_vote(slashable_attestation_1.data, slashable_attestation_2.data)`.
-* Verify that `verify_slashable_attestation(state, slashable_attestation_1)`.
-* Verify that `verify_slashable_attestation(state, slashable_attestation_2)`.
-* Let `slashable_indices = [index for index in slashable_attestation_1.validator_indices if index in slashable_attestation_2.validator_indices and state.validator_registry[index].slashed is False]`.
-* Verify that `len(slashable_indices) >= 1`.
-* Run `slash_validator(state, index)` for each `index` in `slashable_indices`.
+```python
+def process_attester_slashing(attester_slashing: ProposerSlashing,
+                              state: BeaconState):
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Prefer to re-order the parameters: `state` and then `attester_slashing` as most helper functions + fix type hinting.
```python
def process_attester_slashing(state: BeaconState,
                              attester_slashing: ProposerSlashing) -> None:
```

File: specs/core/0_beacon-chain.md [hunk 7]
```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
...
-* Verify that `state.slot < attestation.data.slot + SLOTS_PER_EPOCH.
-* Verify that `attestation.data.justified_epoch` is equal to `state.justified_epoch if slot_to_epoch(attestation.data.slot + 1) >= get_current_epoch(state) else state.previous_justified_epoch`.
-* Verify that `attestation.data.justified_block_root` is equal to `get_block_root(state, get_epoch_start_slot(attestation.data.justified_epoch))`.
-* Verify that either (i) `state.latest_crosslinks[attestation.data.shard] == attestation.data.latest_crosslink` or (ii) `state.latest_crosslinks[attestation.data.shard] == Crosslink(crosslink_data_root=attestation.data.crosslink_data_root, epoch=slot_to_epoch(attestation.data.slot))`.
-* Verify bitfields and aggregate signature:
+For each `attestation` in `block.body.attestations`, run the following function:
 
 ```python
-    assert attestation.custody_bitfield == b'\x00' * len(attestation.custody_bitfield)  # [TO BE REMOVED IN PHASE 1]
+def process_attestation(attestation: Attestation, state: BeaconState):
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Prefer to re-order the parameters: `state` and then `attestation` as most helper functions + fix type hinting:
```suggestion
def process_attestation(state: BeaconState, attestation: Attestation) -> None:
```

File: specs/core/0_beacon-chain.md (hunk 4 above)
Comment (hwwhww) [reply specs/core/0_beacon-chain.md]: Fix type hinting:
```python
def process_proposer_slashing(state: BeaconState,
                              proposer_slashing: ProposerSlashing) -> None:
```

File: specs/core/0_beacon-chain.md [hunk 8]
```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
...
+    assert (
+        # Case 1: Latest crosslink matches the one in the state
+        state.latest_crosslinks[attestation.data.shard] == attestation.data.latest_crosslink or
+        # Case 2: State has already been updated, state's latest crosslink matches the crosslink
+        # the attestation is trying to create
+        state.latest_crosslinks[attestation.data.shard] == Crosslink(
+            crosslink_data_root=attestation.data.crosslink_data_root,
+            epoch=slot_to_epoch(attestation.data.slot)
+        )
+    )
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: I'd like to propose @ralexstokes's refactoring in Trinity:
```python
    acceptable_crosslink_data_roots = {
        attestation.data.latest_crosslink,
        Crosslink(
            crosslink_data_root=attestation.data.crosslink_data_root,
            epoch=slot_to_epoch(attestation.data.slot)
        ),
    }
    assert state.latest_crosslinks[attestation.data.shard] in acceptable_crosslink_data_roots
```

File: specs/core/0_beacon-chain.md [hunk 9]
```diff
@@ -1741,32 +1795,66 @@ process_deposit(state, deposit)
...
-
-* Let `validator = state.validator_registry[exit.validator_index]`.
-* Verify that `validator.exit_epoch > get_delayed_activation_exit_epoch(get_current_epoch(state))`.
-* Verify that `get_current_epoch(state) >= exit.epoch`.
-* Verify that `bls_verify(pubkey=validator.pubkey, message_hash=signed_root(exit, "signature"), signature=exit.signature, domain=get_domain(state.fork, exit.epoch, DOMAIN_EXIT))`.
-* Run `initiate_validator_exit(state, exit.validator_index)`.
+For each `exit` in `block.body.voluntary_exits`, run the following function:
+
+```python
+def process_exit(exit: VoluntaryExit, state: BeaconState):
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Prefer to re-order the parameters: `state` and then `exit` as most helper functions + fix type hinting:
```suggestion
def process_exit(state: BeaconState, exit: VoluntaryExit) -> None:
```

Comment (hwwhww) [reply specs/core/0_beacon-chain.md]: It seems `exit` is a keyword in Python. What do you think about using `voluntary_exit`?

File: specs/core/0_beacon-chain.md [hunk 10]
```diff
@@ -1741,32 +1795,66 @@ process_deposit(state, deposit)
...
-* Verify that `get_current_epoch(state) >= state.validator_registry[transfer.from].withdrawable_epoch` or `state.validator_registry[transfer.from].activation_epoch == FAR_FUTURE_EPOCH`.
-* Verify that `state.validator_registry[transfer.from].withdrawal_credentials == BLS_WITHDRAWAL_PREFIX_BYTE + hash(transfer.pubkey)[1:]`.
-* Verify that `bls_verify(pubkey=transfer.pubkey, message_hash=signed_root(transfer, "signature"), signature=transfer.signature, domain=get_domain(state.fork, slot_to_epoch(transfer.slot), DOMAIN_TRANSFER))`.
-* Set `state.validator_balances[transfer.from] -= transfer.amount + transfer.fee`.
-* Set `state.validator_balances[transfer.to] += transfer.amount`.
-* Set `state.validator_balances[get_beacon_proposer_index(state, state.slot)] += transfer.fee`.
+For each `transfer` in `block.body.transfers`, run the following function:
+
+```python
+def process_exit(transfer: Transfer, state: BeaconState):
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Prefer to re-order the parameters: `state` and then `transfer` as most helper functions + fix type hinting:
```suggestion
def process_exit(state: BeaconState,transfer: Transfer) -> None:
```

**Pull request merged**